sources/parsers/parsers.ml

```diff
@@ -47,37 +47,63 @@ let register_parser ~lang new_parser =
     end;
   parsers := (lang, new_parser) :: !parsers
 
+let debug_no_parser_for_no_extension () =
+  if verbose() then
+    eprintf
+      "error: no extension found, can't infer input format@."
+
 let debug_no_parser_for_extension s =
   if verbose() then
     eprintf
-      "error: no parser registered for ext %S. Using default lang ...@." s
+      "error: no parser registered for ext %S @." s
 
 let debug_no_parser_for_input_format s =
   if verbose() then
     eprintf
       "error: no parser registered for the provided input format %S ?@." s
 
+let get_input_parser () =
+  match Options.input_format () with
+  | Options.Native -> begin
+      try List.assoc ".ae" !parsers
+      with Not_found ->
+        debug_no_parser_for_input_format "ae";
+        exit 1
+    end
+  | Options.Smtlib2 -> begin
+      try List.assoc ".smt2" !parsers
+      with Not_found ->
+        debug_no_parser_for_input_format "smtlib";
+        exit 1
+    end
+  | Options.Why3 -> begin
+      try List.assoc ".why3" !parsers
+      with Not_found ->
+        debug_no_parser_for_input_format "why3";
+        exit 1
+    end
+  | Options.Unknown s -> begin
+      debug_no_parser_for_input_format s;
+      exit 1
+    end
+
 let get_parser lang_opt =
-  let i = Options.input_format () in
   match lang_opt with
-  | Some lang ->
-    begin
-      try List.assoc lang !parsers
-      with Not_found ->
-        debug_no_parser_for_extension lang;
-        try List.assoc i !parsers
+  | Some lang -> begin
+      if Options.infer_input_format () then
+        try List.assoc lang !parsers
         with Not_found ->
-          debug_no_parser_for_input_format i;
+          debug_no_parser_for_extension lang;
           exit 1
+      else
+        get_input_parser ()
     end
-
   | None ->
-    begin
-      try List.assoc i !parsers
-      with Not_found ->
-        debug_no_parser_for_input_format i;
-        exit 1
-    end
+    if Options.infer_input_format () then begin
+      debug_no_parser_for_no_extension ();
+      exit 1 end
+    else
+      get_input_parser ()
```

Comment:
This whole block seems very strange..; the way I read it:

if no language is provided (because we matched `None`), then, if inference is activated, print an error (if verbose, which is a problem, the error should always be printed since we exit right after..., additionally, using exit directly is **BAD**, this should be an exception catched at top-level), else, if no inference is actiavted (and we still don't have a language), try and get the input parser ?

 Either the code is wrong, or there should be a lot more documentation around here to explain what is going on and what this should be doing.